02_knight_and_princess/dana_stepina/main.cpp

```diff
@@ -0,0 +1,97 @@
+#include <array>
+#include <fstream>
+#include <iostream>
+#include <vector>
+
+using std::cout;
+using std::endl;
+
+using VertexId = int;
+using EdgeId = int;
+
+struct Vertex {
+  const VertexId id = 0;
+  std::vector<int> edges_id;
+};
+struct Edge {
+  const EdgeId id = 0;
+  int vertex_start = 0;
+  int vertex_end = 0;
+};
+
+class Graph {
+ public:
+  void add_edges(std::vector<std::array<int, 2>> graph);
+  void add_vertices();
+  const std::vector<Edge>& get_edges() { return edges_; }
+  const std::vector<Vertex>& get_vertices() { return vertices_; }
+
+ private:
+  std::vector<Edge> edges_;
+  std::vector<Vertex> vertices_;
+};
+
+void Graph::add_edges(std::vector<std::array<int, 2>> vertex_connections) {
+  for (int it = 0; it < vertex_connections.size(); it++)
+    edges_.push_back(
+        {it, vertex_connections[it][0], vertex_connections[it][1]});
+}
+void Graph::add_vertices() {
+  int count_vertices = edges_.back().vertex_end;
+  for (VertexId vertex = 0; vertex <= count_vertices; vertex++) {
+    vertices_.push_back({vertex});
+    for (auto edge : edges_) {
+      if ((edge.vertex_start == vertex) || (edge.vertex_end == vertex)) {
+        vertices_[vertex].edges_id.push_back(edge.id);
+      }
+    }
+  }
+}
+
+//Вывод в файл json
+void print_vertex_json(std::ofstream& out, Graph& graph) {

```

Comment:
1. Аргументы нужно передавать как `const` ссылки, иначе ожидается, что они будут меняться внутри функции:
```suggestion
void print_vertex_json(std::ofstream& out, const Graph& graph) {
```
2. "Распечатать вершины" и "Записать данные в файл" - это 2 задачи, а значит функцию можно упростить и избавиться от лишнего аргумента:
```suggestion
std::string print_vertex_json(const Graph& graph) {
```
3. Ответственность этой функции - распечатать вершины. Зачем ей знать о всем графе?
И к тому же `print_vertex` это единственное число и переводится как `распечатать вершину`, мы же печатаем коллекцию вершин, значит должны использовать множественное число:
```suggestion
std::string print_vertices_json(const std::vector<Vertex>& vertices) {
```
Но в таком случае опять, логика усложнена, зачем нам печатать вектор вершин, когда распечатать одну вершину - намного более гибкая и масштабируемая логика:
```suggestion
std::string print_vertex_json(const Vertex& vertex) {
```
Надеюсь тебе понятна данная логическая цепочка?